### PR TITLE
add tranposed gmm tiling for wo

### PR DESCRIPTION
**Pull request by suexu1025**
# Description

Previous all gmms used same tiling configuration. 

This PR enabled same tilings for wi-0, wi-1, and transposed for wo.

This PR improved step time from [41.5s](http://xprof.corp.google.com/trace_viewer/qinwen-6480524894885440759) to [39.5s](http://xprof.corp.google.com/trace_viewer/qinwen-10024501887474315790).

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -1038,22 +1033,32 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
           group_sizes=group_sizes,
           expert_assignments=selected_experts,
       )
-      layer_w0 = gmm_fn(x, w0)
+      w0_tile_size = (
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Could we rename this as `w_i_tile_size` vs. `w_o_tile_size`?

File: src/MaxText/layers/moe.py [hunk 2]
```diff
@@ -1038,22 +1033,32 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
...
+          self.config.tile_weight_dim,
+      )
+      w_out_tile_size = (
+          self.config.tile_batch_seq,
+          self.config.tile_weight_dim,
+          self.config.tile_activation_dim,
+      )
+      layer_w0 = gmm_fn(x, w0,tiling=w0_tile_size)
       if self.get_tensor_transpose_parallelism_size() > 1:
         layer_w0 = jax.lax.psum(layer_w0, "tensor_transpose")
```
[the remark this answers is not in the document]
Comment (suexu1025) [reply src/MaxText/layers/moe.py:1041]: not applicable

File: src/MaxText/layers/moe.py (hunk 1 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: updated, thanks

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -1038,22 +1033,32 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
           group_sizes=group_sizes,
           expert_assignments=selected_experts,
       )
-      layer_w0 = gmm_fn(x, w0)
+      wi_tile_size = (
+          self.config.tile_batch_seq,
+          self.config.tile_activation_dim,
+          self.config.tile_weight_dim,
+      )
+      wo_out_tile_size = (
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Nit: could we apply `wo_tile_size` to align?

File: src/MaxText/layers/moe.py [hunk 4]
```diff
@@ -1038,22 +1033,32 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
...
+          self.config.tile_batch_seq,
+          self.config.tile_weight_dim,
+          self.config.tile_activation_dim,
+      )
+      layer_w0 = gmm_fn(x, w0,tiling=wi_tile_size)
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: I think Gemini mentioned this line :) 

From `layer_w0 = gmm_fn(x, w0,tiling=wi_tile_size)` to `layer_w0 = gmm_fn(x, w0, tiling=wi_tile_size)`

Comment (suexu1025) [reply src/MaxText/layers/moe.py]: I didn't see Gemini comment on this line, but 
on line 1048, which is not on the wrong line?
-layer_w0 = jax.lax.psum(layer_w0, "tensor_transpose")
+layer_w0 = gmm_fn(x, w0, tiling=w0_tile_size)

**Review by gobbleturk — APPROVED** [review]
Thanks qinwen!
